plenum/test/txn_author_agreement/acceptance/test_taa_acceptance_validation.py

```diff
@@ -1,124 +1,224 @@
-        InvalidClientTaaAcceptanceError,
+        validation_error,
         match=(
             "Txn Author Agreement acceptance digest is invalid or non-latest:"
             " provided {}, expected {}"
             .format(
-                calc_taa_digest('some-taa', 'some-taa-version'),
+                taa_digest,
                 calc_taa_digest(latest_taa.text, latest_taa.version)
             )
         )
     ):
-        validate_taa_acceptance(domain_req)
+        validate_taa_acceptance(request_dict)
 
 
 @pytest.mark.taa_acceptance_mechanism('some-unknown-mech')
 def test_taa_acceptance_mechanism_inappropriate(
-    validate_taa_acceptance, domain_req
+    validate_taa_acceptance, validation_error, request_dict
 ):
     with pytest.raises(
-        InvalidClientTaaAcceptanceError,
+        validation_error,
         match=(
             "Txn Author Agreement acceptance mechanism is inappropriate:"
-            " provided {}"
-            .format('some-unknown-mech')
-        )  # TODO more strict error
+            " provided {}".format('some-unknown-mech')
+        )
     ):
-        validate_taa_acceptance(domain_req)
+        validate_taa_acceptance(request_dict)
 
 
 def test_taa_acceptance_time_near_lower_threshold(
-    tconf, validate_taa_acceptance, domain_req, latest_taa
+    tconf, txnPoolNodeSet, validate_taa_acceptance, validation_error,
+    turn_off_freshness_state_update, max_last_accepted_pre_prepare_time,
+    request_dict, latest_taa, monkeypatch
 ):
     taa_ts = latest_taa.txn_time
-    pp_time = taa_ts + 1
+    pp_time = (
+        max_last_accepted_pre_prepare_time +
+        randint(0, tconf.TXN_AUTHOR_AGREEMENT_ACCEPANCE_TIME_AFTER_PP_TIME)
+    )
 
     lower_threshold = taa_ts - tconf.TXN_AUTHOR_AGREEMENT_ACCEPANCE_TIME_BEFORE_TAA_TIME
     upper_threshold = pp_time + tconf.TXN_AUTHOR_AGREEMENT_ACCEPANCE_TIME_AFTER_PP_TIME
 
-    domain_req.taaAcceptance[f.TAA_ACCEPTANCE_TIME.nm] = lower_threshold
-    validate_taa_acceptance(domain_req, pp_time)
+    patch_pp_time(txnPoolNodeSet, monkeypatch, pp_time)
+
+    request_dict[f.TAA_ACCEPTANCE.nm][f.TAA_ACCEPTANCE_TIME.nm] = lower_threshold
+    validate_taa_acceptance(request_dict)
 
-    domain_req.taaAcceptance[f.TAA_ACCEPTANCE_TIME.nm] = lower_threshold - 1
+    request_dict[f.TAA_ACCEPTANCE.nm][f.TAA_ACCEPTANCE_TIME.nm] = lower_threshold - 1
     with pytest.raises(
-        InvalidClientTaaAcceptanceError,
+        validation_error,
         match=(
             r"Txn Author Agreement acceptance time is inappropriate:"
             " provided {}, expected in \[{}, {}\]"
             .format(
-                domain_req.taaAcceptance[f.TAA_ACCEPTANCE_TIME.nm],
+                request_dict[f.TAA_ACCEPTANCE.nm][f.TAA_ACCEPTANCE_TIME.nm],
                 lower_threshold,
                 upper_threshold
             )
         )
     ):
-        validate_taa_acceptance(domain_req, pp_time)
+        validate_taa_acceptance(request_dict)
 
 
```

Comment:
I still don't see an explicit and clear test for the following requirement: pp_time, and not the local node time needs to be used in TAA's acceptance time validation.

I would imagine the following test:
```
def test_taa_acceptance_uses_pp_time_instead_of_current_time(
    tconf, validate_taa_acceptance, domain_req
    validate_taa_acceptance, validation_error, request_dict
):
    pp_time = 10
    now = pp_time + 5 *  tconf.TXN_AUTHOR_AGREEMENT_ACCEPANCE_TIME_AFTER_PP_TIME 
    patch_pp_time(txnPoolNodeSet, monkeypatch, pp_time=pp_time )
    patch_now(txnPoolNodeSet, monkeypatch, now=now)

    request_dict[f.TAA_ACCEPTANCE.nm][f.TAA_ACCEPTANCE_TIME.nm] = pp_time 
    validate_taa_acceptance(request_dict)

    request_dict[f.TAA_ACCEPTANCE.nm][f.TAA_ACCEPTANCE_TIME.nm] = now
    with pytest.raises(validation_error,  match=("Txn Author Agreement acceptance time is inappropriate")):
        validate_taa_acceptance(request_dict)
```

Reply:
added